check/format-incremental

```diff
@@ -35,77 +36,87 @@ cd "$(git rev-parse --show-toplevel)"
+        rev="${base}"
     fi
-fi
-base="$(git merge-base ${rev} HEAD)"
-if [ "$(git rev-parse ${rev})" == "${base}" ]; then
-    echo -e "Comparing against revision '${rev}'." >&2
+
+    # Get the _test version of changed python files.
+    format_files=$(git diff --name-only ${rev} -- | grep "\.py$" | grep -v "_pb2\.py$")
 else
-    echo -e "Comparing against revision '${rev}' (merge base ${base})." >&2
-    rev="${base}"
+    echo -e "Formatting all python files." >&2
+    format_files=$(find . -name "*.py" | grep -v "_pb2\.py$")
 fi
 
-# Get the _test version of changed python files.
-needed_changes=0
-changed_files=$(git diff --name-only ${rev} -- | grep "\.py$" | grep -v "_pb2\.py$")
-esc=$(printf '\033')
-for changed_file in ${changed_files}; do
-    # Extract changed line ranges from diff output.
-    changed_line_ranges=$( \
-        git diff --unified=0 "${rev}" -- "${changed_file}" \
-        | perl -ne 'chomp(); if (/@@ -\d+(,\d+)? \+(\d+)(,)?(\d+)? @@/) {$end=$2+($4 or 1)-1; print "--lines=$2-$end "}' \
-    )
+any_files=0
+for format_file in ${format_files}; do
+  any_files=1
+done
+
+if (( any_files == 0 )); then
+    echo -e "\033[32mNo files to format\033[0m."
+    exit 0
+fi
 
-    if [[ "${changed_line_ranges}" != "--lines=0-0 " ]]; then
-        # Do the formatting.
-        results=$(yapf --style=google --diff "${changed_file}" ${changed_line_ranges})
 
-        # Print colorized error messages.
-        if [ ! -z "${results}" ]; then
-            needed_changes=1
-            if (( only_print == 0 )); then
-                $(yapf --style=google --in-place "${changed_file}" ${changed_line_ranges})
-            else
-                echo -e "\n\033[31mChanges in ${changed_file} require formatting:\033[0m\n${results}" \
-                    | sed "s/^\(+ .*\)$/${esc}[32m\\1${esc}[0m/" \
-                    | sed "s/^\(- .*\)$/${esc}[31m\\1${esc}[0m/"
-            fi
-        fi
-     fi
-done
+args=()
+if (( only_print == 1 )); then
+    args=("--check" "--diff")
+fi
 
-if (( needed_changes == 0 )); then
-    echo -e "\033[32mNo formatting needed on changed lines\033[0m."
-elif (( only_print == 1 )); then
-    echo -e "\033[31mSome formatting needed on changed lines\033[0m."
-    exit 1
-else
-    echo -e "\033[33mReformatted changed lines\033[0m."
+# Workaround bug in black: https://github.com/psf/black/issues/1629
+# Once that is fixed upstream, we can just do:
+# black --color "${args[@]}" ${format_files}
+# exit $?
+LOGS=$(black --color "${args[@]}" ${format_files} 2>&1)
+STATUS=$?
+echo "$LOGS"
+if [[ "$STATUS" == "123" && "$LOGS" =~ ^.*"INTERNAL ERROR: Black produced different code on the second pass of the formatter.".*$ ]]; then
+  echo -e "\033[31mWarning, it seems we ran into https://github.com/psf/black/issues/1629. Falling back to calling black --fast twice. Please double check that your code is still valid!\033[0m"
+  black --fast --color "${args[@]}" ${format_files}
+  STATUS=$?
```

Comment:
I think we can be conservative here and exit on failure. Reason: we know that for the 1629 bug the first formatting succeeds. Thus we can be specific about it here and ask the user to file a bug instead if the first command fails.